### PR TITLE
Fix the wrong result bug when casting string as datetime with time zone or illegal chars

### LINKED ISSUE #8754: cast string as datetime with time zone results in loss of precision
## Bug Report

Please answer these questions before submitting your issue. Thanks!
related to https://github.com/pingcap/tidb/issues/49555

### 1. Minimal reproduce step (Required)
```
use test;
drop table if exists t;
create table t(a varchar(50));
alter table test.t set tiflash replica 1;
insert into t values ('2020-01-01 12:00:00.123456 +0600 PST');
insert into t values ('2020-01-01 12:00:00.123456 -0600 PST');
set tidb_allow_mpp=1;set tidb_enforce_mpp=1; set tidb_isolation_read_engines='tiflash';
select cast(a as datetime(3)) from t;
```

### 2. What did you expect to see? (Required)
```
+-------------------------+
| cast(a as datetime(3))  |
+-------------------------+
| 2020-01-01 12:00:00.123 |
| 2020-01-01 12:00:00.123 |
+-------------------------+
```
### 3. What did you see instead (Required)
```
+-------------------------+
| cast(a as datetime(3))  |
+-------------------------+
| 2020-01-01 12:00:00.000 |
| 2020-01-01 12:00:00.000 |
+-------------------------+
```
### 4. What is your TiFlash version? (Required)
v8.0.0-alpha-126-gc0476a0

### PR DESCRIPTION
**Pull request by solotzg**
### What problem does this PR solve?

Issue Number: close #8754

Problem Summary:

### What is changed and how it works?

```commit-message
Fix the wrong result bug when casting string as datetime with time zone or illegal chars
- func `getFracIndex` ignore time zone.
- func `parseFrac` ignore illegal suffix after microsecond format.
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix the wrong result bug when casting string as datetime with time zone or illegal chars
```

### CODE REVIEW

**Review by guo-shaoge — APPROVED** [review]

File: dbms/src/Common/MyTime.cpp [hunk 1]
```diff
@@ -745,7 +745,7 @@ inline bool numberToDateTime(Int64 number, MyDateTime & result, bool allowZeroDa
 }
 
 // returns frac, overflow, matched. eg., "999" fsp=2 will overflow.
-std::tuple<UInt32, bool, bool> parseFrac(const String & str, int8_t fsp)
+std::tuple<UInt32, bool, bool> parseFrac(const std::string_view str, int8_t fsp)
```
Comment (windtalker) [inline dbms/src/Common/MyTime.cpp:748]: The caller don't need to be updated after the input argument type is changed?

Comment (solotzg) [reply dbms/src/Common/MyTime.cpp:748]: In this case, `const String &` and `std::string_view` have the same effect. The caller won't change the original string buffer.

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**